### ISSUE TITLE
I hope it can support more program languages

### ISSUE THREAD

**Issue by Rickdroid** [SaekiRaku/vscode-rainbow-fart#2]
I hope it can support more program languages, such as golang, C++. I always write program with golang. I need it makes me BOKI!!

**Comment by Rickdroid** [comment]
python also be ok

**Comment by SaekiRaku** [comment]
Thanks for your feedback.

Diversity is the primary goal of this project, I hope it can suitable for more programmer no matter what program languages they are using. However, it needs a lot of works to do and take a lot of time. The built-in voices of the extension were came from a friend of mine, two of us certainly can not handle all of these requests in the future.

So, I'm hoping the community can contribute more voice packages. And speaking of diversity, it's not only mean different program languages, but also mean different gender, country, etc. Therefore, almost everyone can contribute, I'll be happy to accept such a PR.

In general, I think the next move of this project may be like this:
* Add CONTRIBUTE.md so that people who willing to help can contribute correctly.
* Make the [Customize Voice Package](https://saekiraku.github.io/vscode-rainbow-fart/#/zh/voice-packages.md) document more perfect.
* Build a tool that let people add voice easily so that some non-programmer can also create voice packages.

**Comment by Rickdroid** [comment]
Thx

> Thanks for your feedback.
> 
> Diversity is the primary goal of this project, I hope it can suitable for more programmer no matter what program languages they are using. However, it needs a lot of works to do and take a lot of time. The built-in voices of the extension were came from a friend of mine, two of us certainly can not handle all of these requests in the future.
> 
> So, I'm hoping the community can contribute more voice packages. And speaking of diversity, it's not only mean different program languages, but also mean different gender, country, etc. Therefore, almost everyone can contribute, I'll be happy to accept such a PR.
> 
> In general, I think the next move of this project may be like this:
> 
> * Add CONTRIBUTE.md so that people who willing to help can contribute correctly.
> * Make the [Customize Voice Package](https://saekiraku.github.io/vscode-rainbow-fart/#/zh/voice-packages.md) document more perfect.
> * Build a tool that let people add voice easily so that some non-programmer can also create voice packages.

thx for your reply, I hope your project will more and more NB(in Chinese).

**Comment by fengyuentau** [comment]
@SaekiRaku Could you provide some guides on how to support other language if you are too busy to deal with all the requests. Though I am not a JS expert, I am willing to hack into this repo.

**Comment by SaekiRaku** [comment]
@fengyuentau Thanks for your attention, here is a simple guide on customizing voice packages:  https://saekiraku.github.io/vscode-rainbow-fart/#/zh/voice-packages.md If you have any question or improvement about this document, I'm glad to have your feedback.

P.S. The English version of this document was not translated,  I'll be happy to accept PR.

Thanks again.

**Comment by fengyuentau** [comment]
@fengyuentau Sorry I did not make myself understood. To clarify, I want it to support more program language since I am not a JS programmer. But the page you provide also works, I will try modify it to fit my case. Thanks.

**Comment by chenzhekl** [comment]
IIRC, in the current implementation, voice is trigged based on the keyword that is being typed. It would be great if the keyword list can be provided separately per the programming language. I guess that's what the issue is asking for.

**Comment by SaekiRaku** [comment]
@chenzhekl 

Sorry, I'm still confused. 

In the current implementation, if you want to make the voice support other programming languages, you can simply add more keywords to it. Like this:

```js
// manifest.json - contributes
{
    // JavaScript, Golang
    "keywords": ["function", "func"],
    "voices": [ "function_01.mp3"  ]
}
```

Are you suggesting to make the json file be like this?

```js
// manifest.javascript.json
{
    // JavaScript
    "keywords": ["function"],
    "voices": [ "function_01.mp3"  ]
}

// manifest.golang.json
{
    "keywords": ["func"],
    "voices": [ "function_01.mp3"  ]
}
```

**Comment by fengyuentau** [comment]
@SaekiRaku I guess that's the key point of this issue. I tried to edit `~/.vscode/extension/vscode-rainbow-fart/src/built-in-voice-packages/built-in-voice-chinese/contributes.json` to add more keywords, but it does not work. The webpage still shows the old keywords. Do you have any ideas?

**Comment by chenzhekl** [comment]
@SaekiRaku Thanks for the explanation! My bad. It's great to know that multi-language is already supported. Would you mind adding some examples to the documentation? I believe that can benefit more people. Anyway, thank you very much for this interesting extension!

**Comment by chenzhekl** [comment]
@SaekiRaku Yeah, that's what we're hoping for. For one, using a separate keyword list could greatly ease the maintenance. Moreover, that might help resolve ambiguities between different programming languages. For example, the fat arrow (=>) is used to define a lambda in C# and ES6 while in Rust it's used for the pattern matching. Of course, using multiple keyword lists could bring in some duplications. A compromise might be providing a global keyword list plus a language-specific one. What do you think?

**Comment by SaekiRaku** [comment]
@fengyuentau 

The `contributes.json` will be cached after extension activated, you need to click the `refresh` button on the web to reload it. But if the built-in voice package has changed in future updates, it will re-copy all files to the built-in voice packages directory, then you will lose your changes.

I recommend you to copy the built-in voice package and change the directory name and the `name` in `manifest.json`, then add the keywords. At last, click `refresh`.

**Comment by fengyuentau** [comment]
@SaekiRaku I followed you suggestion, but it seems that the `refresh` does not work. Neither it refreshed the keywords that I changed to the built-in package, nor did it refresh the new directory that I copied and modified. I installed this extension on my remote server, is it the problem? I can type existed keywords and hear the voice played.

**Comment by SaekiRaku** [comment]
@fengyuentau 

I've tried to use the extension with VSCode Remote on a CentOS server, it works for me. Here is my step:

1. Install the extension on remote server and enable it.
2. Copy and modify the `contributes.json` on the remote server, the path should be like this: `/$USER/.vscode-server/data/User/globalStorage/saekiraku.rainbow-fart/voice-packages/built-in-voice-chinese/contributes.json`
3. click `refresh`

**DO NOT EDIT** `/$USER/.vscode-server/extensions/saekiraku.rainbow-fart-1.2.0/src/built-in-voice-packages`, it won't work.

If you are using windows server, I'm not sure about the path. You can clone this project, and run VSCode command: `> Debug: Start Debugging`. It will bring up a new VSCode window, use that to connect to the remote server and enable the extension. You will see the path in the `DEBUG CONSOLE` panel.

**Comment by SaekiRaku** [comment]
@chenzhekl 

I've conceived some methods to achieve the request, too. for example:

```
{
    "keywords": ["=>"],
    "include": [ "rust" ],
    "voices": [ "pattern_matching.mp3" ]
},
{
    "keywords": ["=>"],
    "exclude": [ "rust" ],
    "voices": [ "lambda.mp3" ]
}
```

But I think there still has a problem. See, I didn't know the meaning of fat arrow in Rust, and maybe there has another language that also has a different meaning for the fat arrow and we don't know. It will be an annoying job to keep maintain the language-specific keywords list or include/exclude list, and it will keep users waiting for the changing. So I think maybe the best way is to provide a function in the web UI to let users disable specific keywords/voices.

**Comment by chenzhekl** [comment]
@SaekiRaku True, that's indeed a problem. I don't have much knowledge about how the voice package is structured, though. I'm wondering if it's possible to bundle the configuration file with the voice resources and make such a voice package the smallest distribution unit (as a separate extension). Then maintaining the keyword list becomes the job of the voice package maintainer. Besides, the user is able to define custom keywords on top of the pre-defined keywords to extend existing voice packages.

P.S. [inter-extension communication](https://stackoverflow.com/questions/50058517/how-to-communicate-between-vscode-extensions)

**Comment by PM-Ops** [comment]
`自我的一个朋友` ---> 无中生友，真的只是朋友咩？

![image](https://user-images.githubusercontent.com/62615415/85353364-9f167080-b53a-11ea-95c2-5f2cf5626bce.png)

**Comment by SaekiRaku** [comment]
@PM-Ops 

![IMG_5930](https://user-images.githubusercontent.com/12109558/85353761-9bcfb480-b53b-11ea-8a08-90a617168c62.JPG)

**Comment by stardiviner** [comment]
@SaekiRaku 嗨，用了自己开发的emacs-rainbow-fart一段时间了，有用户报告说缺少class语音，不知道作者朋友能不能添加下。让默认语音包更加完善点。非常感谢 😄
